KAFKA-5445: Document exceptions thrown by AdminClient methods

adyach opened the pull request:
Exceptions are processed internally in KafkaAdminClient without throwing them to the client code, hence the documentation of the exception is done in unusual way.

adyach committed:
kafka-5445: create topic exception description

adyach committed:
kafka-5445: rest of the commands descritpion

adyach commented:
@ijuma could you take a look at it, please?

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -46,6 +48,8 @@
      * @param props The configuration.
      * @return The new KafkaAdminClient.
+     * @throws ConfigException
+     * @throws KafkaException
```
We should probably skip adding these unless we're going to say more about them?

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -229,6 +317,18 @@ public DescribeAclsResult describeAcls(AclBindingFilter filter) {
+     *     <li>{@link org.apache.kafka.common.errors.TimeoutException}
+     *     if the request was not completed in within the given {@link DescribeAclsOptions#timeoutMs()}.
+     *     </li>
+     * </ul>
```
Should add InvalidRequestException

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -256,6 +356,21 @@ public CreateAclsResult createAcls(Collection<AclBinding> acls) {
+     *     <li>{@link org.apache.kafka.common.errors.ClusterAuthorizationException}
+     *     if cluster authorization failed</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidRequestException}
+     *     if invalid resource name was provided</li>
```
This is only one reason why the request might be invalid.  Probably best to just say "if the request was invalid" -- the specific exception text explains more about why it was invalid.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -280,6 +395,21 @@ public DeleteAclsResult deleteAcls(Collection<AclBindingFilter> filters) {
+     *     <li>{@link org.apache.kafka.common.errors.ClusterAuthorizationException}
+     *     if cluster authorization failed</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidRequestException}
+     *     if invalid resource name was provided</li>
```
same comment here.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -229,6 +317,18 @@ public DescribeAclsResult describeAcls(AclBindingFilter filter) {
+     * {@link DescribeAclsResult#values() values()} method of the  returned {@code DescribeAclsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.SecurityDisabledException}
+     *     if security features are disabled</li>
```
It might be better to be more specific here.  "if no authorizer is configured on the broker"

adyach committed:
KAFKA-5445: fixed wording

adyach commented:
@cmccabe thank you for review, I have addressed your comments

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 No test results found.
--none--

adyach committed:
KAFKA-5445: revert code change

asfgit commented:
FAILURE
 8067 tests run, 5 skipped, 2 failed.
--none--

asfgit commented:
FAILURE
 7959 tests run, 5 skipped, 1 failed.
--none--

asfgit commented:
SUCCESS 
 8067 tests run, 5 skipped, 0 failed.
--none--

mjsax commented:
@cmccabe What is the status of this PR?

Retest this please.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -110,6 +110,30 @@ public CreateTopicsResult createTopics(Collection<NewTopic> newTopics) {
+     *     <li>{@link org.apache.kafka.common.errors.InvalidPartitionsException}
+     *     if number of partitions is less than 1</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidReplicationFactorException}
+     *     if replication factor is less than 1 or large than number of available brokers</li>
```
Should be larger

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -110,6 +110,30 @@ public CreateTopicsResult createTopics(Collection<NewTopic> newTopics) {
+     *     <li>{@link org.apache.kafka.common.errors.TopicExistsException}
+     *     if topic name already exists</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidTopicException}
+     *     if topic name collides with other topic name</li>
```
`InvalidTopicException` indicates that the topic name itself is invalid, not that it collides with another topic name

adyach replied:
I found it in the source https://github.com/apache/kafka/blob/88307657dff0202804447d06eef6ecdc808590d8/core/src/main/scala/kafka/admin/AdminUtils.scala#L492

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -110,6 +110,30 @@ public CreateTopicsResult createTopics(Collection<NewTopic> newTopics) {
+     *     <li>{@link org.apache.kafka.common.errors.InvalidTopicException}
+     *     if topic name collides with other topic name</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidReplicaAssignmentException}
+     *     if some of the partitions have different number of replicas or duplicate replica assignment found</li>
```
How about "if the proposed replica assignment is invalid.  For example if some of the partitions have different number of replicas or a duplicate replica assignment was found"

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -110,6 +110,30 @@ public CreateTopicsResult createTopics(Collection<NewTopic> newTopics) {
+     *     <li>{@link org.apache.kafka.common.errors.InvalidReplicaAssignmentException}
+     *     if some of the partitions have different number of replicas or duplicate replica assignment found</li>
+     *     <li>{@link org.apache.kafka.common.errors.PolicyViolationException}
+     *     if the request parameters do not satisfy the user defined policy</li>
```
Might want to emphasize "the policy configured on the broker"

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -162,6 +200,24 @@ public ListTopicsResult listTopics() {
+     *
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.UnknownTopicOrPartitionException}
+     *     if the authorized topic was in metadata first, but was not found during later look up</li>
```
We don't use UnknownTopicOrPartitionException in listTopics

adyach replied:
When I was doing the research, it was there https://github.com/adyach/kafka/blob/KAFKA-5445/core/src/main/scala/kafka/server/KafkaApis.scala#L952

adyach replied:
but it is in https://github.com/apache/kafka/blob/88307657dff0202804447d06eef6ecdc808590d8/core/src/main/scala/kafka/server/KafkaApis.scala#L935

cmccabe replied:
That line is not there any more.

In any case, it doesn't really make sense to return this to the user.  list topics is supposed to list all topics.  What would returning UnknownTopicOrPartition mean in this context?

adyach replied:
@cmccabe I have removed from the doc 7a42286f1, but it is still in the code, so it could be returned to the user of that class.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -162,6 +200,24 @@ public ListTopicsResult listTopics() {
+     *     <li>{@link org.apache.kafka.common.errors.UnknownTopicOrPartitionException}
+     *     if the authorized topic was in metadata first, but was not found during later look up</li>
+     *     <li>{@link org.apache.kafka.common.errors.LeaderNotAvailableException}
+     *     if leader is unavailable</li>
```
We don't use LeaderNotAvailableException in listTopics

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -162,6 +200,24 @@ public ListTopicsResult listTopics() {
+     *     <li>{@link org.apache.kafka.common.errors.LeaderNotAvailableException}
+     *     if leader is unavailable</li>
+     *     <li>{@link org.apache.kafka.common.errors.ReplicaNotAvailableException}
+     *     if the replica is not available</li>
```
We don't use ReplicaNotAvailableException in listTopics

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -162,6 +200,24 @@ public ListTopicsResult listTopics() {
+     *     <li>{@link org.apache.kafka.common.errors.ReplicaNotAvailableException}
+     *     if the replica is not available</li>
+     *     <li>{@link org.apache.kafka.common.errors.TopicAuthorizationException}
+     *     if the authenticated user is not authorized to alter the topic</li>
```
We don't use TopicAuthorizationException in listTopics

adyach replied:
I see in the source https://github.com/apache/kafka/blob/88307657dff0202804447d06eef6ecdc808590d8/core/src/main/scala/kafka/server/KafkaApis.scala#L996

cmccabe commented:
@mjsax : thank you for the reminder.  I re-reviewed this.

Since every future can throw TimeoutException and AuthorizationException, maybe we should document these in the class JavaDoc, rather than repeating these exceptions on every function's JavaDoc.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -183,6 +239,22 @@ public DescribeTopicsResult describeTopics(Collection<String> topicNames) {
+     *     <li>{@link org.apache.kafka.common.errors.LeaderNotAvailableException}
+     *     if leader is unavailable</li>
+     *     <li>{@link org.apache.kafka.common.errors.ReplicaNotAvailableException}
+     *     if the replica is not available</li>
```
not used here

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -183,6 +239,22 @@ public DescribeTopicsResult describeTopics(Collection<String> topicNames) {
+     *     <li>{@link org.apache.kafka.common.errors.UnknownTopicOrPartitionException}
+     *     if the authorized topic was in metadata first, but was not found during later look up</li>
+     *     <li>{@link org.apache.kafka.common.errors.LeaderNotAvailableException}
+     *     if leader is unavailable</li>
```
not used here

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -183,6 +239,22 @@ public DescribeTopicsResult describeTopics(Collection<String> topicNames) {
+     * {@link DescribeTopicsResult#values() values()} method of the  returned {@code DescribeTopicsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.UnknownTopicOrPartitionException}
+     *     if the authorized topic was in metadata first, but was not found during later look up</li>
```
not used here (InvalidTopicException is used instead)

adyach replied:
The request is the same for listTopics, so it is the same here

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -183,6 +239,22 @@ public DescribeTopicsResult describeTopics(Collection<String> topicNames) {
+     *     <li>{@link org.apache.kafka.common.errors.ReplicaNotAvailableException}
+     *     if the replica is not available</li>
+     *     <li>{@link org.apache.kafka.common.errors.TopicAuthorizationException}
+     *     if the authenticated user is not authorized to alter the topic</li>
```
Not used here

adyach replied:
The request is the same for listTopics, so it is the same here

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -314,6 +438,25 @@ public DescribeConfigsResult describeConfigs(Collection<ConfigResource> resource
+     * {@code DescribeConfigsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.SecurityDisabledException}
+     *     if security features are disabled</li>
```
Not used here

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -314,6 +438,25 @@ public DescribeConfigsResult describeConfigs(Collection<ConfigResource> resource
+     *     <li>{@link org.apache.kafka.common.errors.InvalidRequestException}
+     *     if unsupported resource type or unexpected broker id</li>
+     *     <li>{@link org.apache.kafka.common.errors.InvalidTopicException}
+     *     if topic name is illegal</li>
```
I don't think this is used here?

adyach replied:
yes, it was mistake

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -341,6 +484,25 @@ public AlterConfigsResult alterConfigs(Map<ConfigResource, Config> configs) {
+     * {@code AlterConfigsResult}</p>
+     * <ul>
+     *     <li>{@link org.apache.kafka.common.errors.SecurityDisabledException}
+     *     if security features are disabled</li>
```
Not used here

mjsax commented:
Thanks for reviewing @cmccabe 
I am fine either way about TimeoutException and AuthorizationException

@adyach The code freeze deadline for 1.1 release is 13/2 -- would be great if we could get this in.

adyach committed:
KAFKA-5445: addressed code review comments

adyach commented:
@cmccabe  @mjsax I've addressed your comments and decided to leave TimeoutException and AuthorizationException per method description, because I see that not all methods throws both of them

mjsax commented:
@adyach Thanks for updating.
@cmccabe Can you review again? RC-0 for `1.1` was not cut yet, so we could still get it into the release.

adyach committed:
aruha-5445: removed UnknownTopicOrPartition

cmccabe commented:
> I've addressed your comments and decided to leave TimeoutException and AuthorizationException per method description, because I see that not all methods throws both of them

@adyach, any method which involves talking to a remote server could time out, if that server takes too long to respond.  Similarly, if we can't get authorization to get the cluster metadata (SASL misconfiguration), we should get `AuthorizationException`, no matter what call we are trying to make.  Since all the admin client methods involve talking to remote servers, they all should be able to get both of these exceptions.  Those exceptions come from the event loop, not from the per-call code.

So, I think we should have a common description for those exceptions... perhaps in the class Javadoc.  We should also document UnknownServerException, since this is the catch-all exception that the user gets if there is an unexpected internal error in the broker.

> @cmccabe Can you review again? RC-0 for 1.1 was not cut yet, so we could still get it into the release.

I understand the desire to get this out, but it's not good to rush.  I already see some divergences in what exceptions can be thrown in trunk versus 1.1, which I think will make backporting take some work.

This PR is difficult to review because it covers so many different calls.  We also need to be really careful that the exceptions we are documenting are real, and can actually be thrown in the calls we're looking at.  This will require looking through the broker code carefully.  (For example, listTopics should not throw UnknownPartitionException, since the MetadataRequest it does is for all topics, not a specific set of topics.)

How about splitting this into smaller PRs?
One PR for createTopics / deleteTopics / listTopics / describeTopics.
Another PR for the ACLs calls.
Another PR for the configs calls.
Another PR for the log dirs calls and whatever is left.

mjsax commented:
@cmccabe I agree it's not worth rushing it in...

mjsax commented:
@adyach @cmccabe We are getting close to 2.0 release -- might be worth to try to get it into this one :)

adyach commented:
@mjsax I will try to split it on smaller PRs as it was suggested by @cmccabe

rajinisivaram commented:
@adyach What is the status of this PR? Are you planning to update to include in 2.0?

adyach commented:
@rajinisivaram I will continue my work here, I am not certain I will have time to finish it in 2.0.0 though.

mjsax commented:
What is the status of this PR?

adyach commented:
@mjsax @cmccabe I have split the PR into the first one https://github.com/apache/kafka/pull/5558
